diff --git a/virttest/qemu_vm.py b/virttest/qemu_vm.py
index f8fee96..5b4f8ee 100644
--- a/virttest/qemu_vm.py
+++ b/virttest/qemu_vm.py
@@ -2367,6 +2367,13 @@ class VM(virt_vm.BaseVM):
             if ats:
                 add_virtio_option("ats", ats, devices, device, dev_type)
 
+        #wq
+        pr_manager_helper = params.get("pr_manager_helper")
+        if pr_manager_helper:
+            cmd = "-object pr-manager-helper,id=%s,path=/var/run/qemu-pr-helper.sock"
+            devices.insert(StrDev("pr_manager_%s" % pr_manager_helper,
+                          cmdline=cmd % pr_manager_helper))
+
         # Add extra root_port at the end of the command line only if there is
         # free slot on pci.0, discarding them otherwise
         func_0_addr = None
